### PR TITLE
[pvr] ability to sort channels and epg by the backend order of channels

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Add the ability to sort channels and epg by the backend order of channels. Instead of extending the API I just used the creation order for the value or sort by.

Resolves issue: https://github.com/xbmc/xbmc/issues/16546

## Motivation and Context

Allows the preservation of order of the backend pvr if that system does not use channel number based ordering.

## How Has This Been Tested?

Using pvr.vuplus and pvr.iptvsimple.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/pvr/channels/PVRChannel.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.cpp]: This is completely wrong as it breaks the object hierarchy. Channel must not depend on any of the channel containers.

File: xbmc/pvr/channels/PVRChannel.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.cpp]: Clear indicator of the broken design. :-(

**Review by ksooo — CHANGES_REQUESTED** [review]
Sorry, did not went through the whole changes. -1 for two reasons: 

1. implementation breaks PVR object call hierarchy. CPVRChannel and CPVREpgInfoTag must not depend on any of the channel containers. 

2. API needs to be extended to transport the order. One should not rely on creation order.

File: xbmc/pvr/channels/PVRChannelGroup.h
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelGroup.h]: in Kodi core we always use a blank after `//`.

File: xbmc/pvr/epg/EpgInfoTag.cpp
Comment (ksooo) [inline xbmc/pvr/epg/EpgInfoTag.cpp]: EPG must not depend on channels. This can lead to deadlocks.

Recordings -> Timers -> Channels -> EPG is the general call order.

**Comment by ksooo** [comment]
If you need to transfer any channel information to CPVREPGInfoTag, do this via its CPVREpgChannelData member.

If you need the order at the CPVRChannel, introduce a member m_iOrder (or the like) at CPVRChanenl.

**Comment by ksooo** [comment]
pvr-component-deps

**Comment by phunkyfish** [comment]
Ok, for point 2) I know what to do.

However for point 1) it’s a bit trickier as both Channel and EpgInfoTag would need to know the current group member that they pertain to without breaking the design. I’ll have to think on that.

**Comment by phunkyfish** [comment]
@ksooo, should now satisfy both rejection points.

File: xbmc/pvr/channels/PVRChannel.h
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.h]: (void) -> ()

File: xbmc/utils/DatabaseUtils.h
Comment (ksooo) [inline xbmc/utils/DatabaseUtils.h]: This must be added before FieldMax

File: xbmc/pvr/windows/GUIViewStatePVR.cpp
Comment (ksooo) [inline xbmc/pvr/windows/GUIViewStatePVR.cpp]: This sort method should only be available if "use backend sort order" is activated in settings.

I would add this sort method at the end, not at the start - given that this is  a very special sort method, compared to the other ones.

File: xbmc/pvr/windows/GUIViewStatePVR.cpp
Comment (ksooo) [inline xbmc/pvr/windows/GUIViewStatePVR.cpp]: This sort method should only be available if "use backend sort order" is activated in settings.

I would add this sort method at the end, not at the start - given that this is  a very special sort method, compared to the other ones.

File: xbmc/utils/SortUtils.cpp
Comment (ksooo) [inline xbmc/utils/SortUtils.cpp]: one space after the comma is enough.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: Please increase minor of ADDON_INSTANCE_VERSION_PVR

=> https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L102

As we stay ABI compatible, no need to increase ADDON_INSTANCE_VERSION_PVR_MIN.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: typo: ... to indicate this channel...

File: xbmc/pvr/PVRDatabase.cpp
Comment (ksooo) [inline xbmc/pvr/PVRDatabase.cpp]: wrong indentation (before, it was right).

File: xbmc/pvr/PVRDatabase.cpp
Comment (ksooo) [inline xbmc/pvr/PVRDatabase.cpp]: wrong indentation (before, it was right).

File: xbmc/pvr/channels/PVRChannel.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.cpp]: coding guideline: for new code, only one blank between var name and =

File: xbmc/pvr/epg/EpgChannelData.h
Comment (ksooo) [inline xbmc/pvr/epg/EpgChannelData.h]: m_iOrder = 0;

File: xbmc/pvr/epg/EpgChannelData.cpp
Comment (ksooo) [inline xbmc/pvr/epg/EpgChannelData.cpp]: As this method only assign the value to the corresponding member, a simple `m_iOrder(channel.ClientOrder())` in the ctor init list is sufficient here.

File: xbmc/pvr/channels/PVRChannelGroup.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelGroup.cpp]: Updating epg channeldata should be done in the implementation of `CPVRChannel::SetClientOrder`, following the other implementations, example: `CPVRChannel::SetChannelNumber`=> https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/channels/PVRChannel.cpp#L652

**Review by ksooo — COMMENTED** [review]
Generally okay now. Needs a bit more love, though.

**Comment by phunkyfish** [comment]
Jenkins build this please

File: xbmc/pvr/channels/PVRChannel.h
Comment (phunkyfish) [reply xbmc/pvr/channels/PVRChannel.h]: Ah, ok. Was sticking with the what was used in the rest of file.

File: xbmc/pvr/windows/GUIViewStatePVR.cpp
Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: It being at the end of the list makes sense but it's not dependent on "use backend sort order". You may like the Kodi numbering but still want the backend order.

Besides, "use backend sort order" is sorting by the backend numbers so it relates to something different. The text for "use backend sort order" could be changed to make it's meaning more clear.

File: xbmc/pvr/channels/PVRChannel.h
Comment (ksooo) [reply xbmc/pvr/channels/PVRChannel.h]: For new code we apply current coding guidelines, regardless of the format of the existing code.

File: xbmc/pvr/windows/GUIViewStatePVR.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: > It being at the end of the list makes sense but it's not dependent on "use backend sort order". You may like the Kodi numbering but still want the backend order.

 Good point.

> Besides, "use backend sort order" is sorting by the backend numbers 

 ... because currently we do not have any other backend order than the numbers. With your PR this changes and thus renaming the setting label could make sense, yes.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: BTW: What's the behavior if I sort by backend order and backend supplies iPosition == 0 for all channels?

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Actually, I just checked and "use backend order" appears to have no effect whatsoever. I guess the left nav sorting was added after this.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: can of worms?

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Maybe should remove "Use backend order" from settings.

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Many worms...

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: The more I think about this, the more I ask myself whether the feature introduced by this PR makes sense. ;-)

Why is it not enough to use the backend channel numbers as sort criteria? Why do we need something "overloading" the numbers. What's the benefit/use case in assigning a channel the number "1" but not wanting to see it at the beginning of channel list, but somewhere in the middle or elsewhere? Isn't the (sole) purpose of channel numbers to order the channels to match users preference? So confusing...

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: If order is zero for all channels that sort does nothing.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Have you seen my question re general sense of this PR?

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: I have my difficulties in seeing the use case for this feature.

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: I think it does make sense. My wife has her favourites group on our PVR, morning channels at the top, evening channels at the bottom.

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: You can also order channels by name and last watched.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Fair enough.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: > Maybe should remove "Use backend order" from settings.

 Why?

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Because the setting doesn't appear to have any effect. PVRChannelGroup sorts the channels, but then they are sorted again by the left nav sort. Or am I missing something...

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: > Because the setting doesn't appear to have any effect.

 The question is: should we fix that?

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: But as you said, can of worms ;) Maybe best to not touch until a rewrite.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: Wait: The purpose of the setting is not to define the order of the channels on screen windows. Its the purpose to respect the backend sort order (defined be the backend channel numbers) when creating local channel numbers!

Backend:
channel A, number 1
channel B, number 5
channel C, number 9

Kodi: use backend sort order active, use backend channel numbers not active
channel A, number 1
channel B, number 2
channel C, number 3

Activating use channel backend numbers implies use backend channel order. The latter gets changed by your PR.

Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: I see. That makes sense.

This PR only changes the order on the screen though. It's doesn't effect the sorting within PVRChannelGroup so should not change the latter.

Comment (ksooo) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: To conclude:

* your feature makes sense
* settings "use backend channel order" makes sense

;-)

File: xbmc/pvr/channels/PVRChannel.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.cpp]: nitpick: maybe we can put the new member after the last existing member (and add a blank line to visually separate methods from members.)

File: xbmc/pvr/windows/GUIViewStatePVR.cpp
Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: I just spotted something, if you set a different channel number per group for the same channel it has no effect. That’s ones for a different PR for different day 😉

File: xbmc/pvr/channels/PVRChannelGroup.h
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelGroup.h]: I'm not a native English speaker, but is "it's" correct here? I thought it has to read "its", no?

File: xbmc/pvr/windows/GUIViewStatePVR.cpp
Comment (phunkyfish) [reply xbmc/pvr/windows/GUIViewStatePVR.cpp]: I completely agree with your summary!!!

File: xbmc/pvr/channels/PVRChannelGroup.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelGroup.cpp]: Do you think you could left align all the & here and in every other affected line you touched to make these lines obey our coding guidelines?

**Review by ksooo — COMMENTED** [review]
Let the nitpicking begin...

File: xbmc/pvr/channels/PVRChannelGroup.h
Comment (phunkyfish) [reply xbmc/pvr/channels/PVRChannelGroup.h]: Yes, you are correct.

File: xbmc/pvr/channels/PVRChannelGroup.cpp
Comment (phunkyfish) [reply xbmc/pvr/channels/PVRChannelGroup.cpp]: Done

**Comment by ksooo** [comment]
> [pvr] Allow use genre strings to be set from subtype so genre type (c…

Commit message seems unrelated to this PR. ;-)

**Comment by phunkyfish** [comment]
> Commit message seems unrelated to this PR. ;-)

Ya, ran the wrong history command, fixed now.

**Comment by ksooo** [comment]
So, does it still work after all the review induced changes?

**Comment by phunkyfish** [comment]
Yes, why do you think the changes would effect it working?

**Comment by ksooo** [comment]
> Did you spot something?

 No. I was just asking.

**Comment by phunkyfish** [comment]
Cool, tested with iptvsimple and with also with vuplus. Works with both.

**Comment by ksooo** [comment]
From a user's point of view, what would you expect to happen, if you select "Backend order" if the pvr client addon does not fill the new iOrigin field? Nothing? Hmmm... Shouldn't we add a fallback logic to sort by backend channel numbers, then?

**Comment by phunkyfish** [comment]
Yes, that makes sense, agreed. That's the missing piece.

**Comment by phunkyfish** [comment]
It might be simpler in UpdateClientOrder to set it to the backend channel number if iOrder is 0.

**Comment by phunkyfish** [comment]
Done.

File: xbmc/pvr/channels/PVRChannel.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.cpp]: One semicolon ought to be sufficient. ;-)

**Review by ksooo — COMMENTED** [review]
Almost done, I'd say.

File: xbmc/pvr/channels/PVRChannel.cpp
Comment (phunkyfish) [reply xbmc/pvr/channels/PVRChannel.cpp]: :(o

**Comment by phunkyfish** [comment]
Question. The iOrder field send from the addon, if this is spent from an addon on on version 6.0.0 it won't be sent. Is it possible to ensure that in that case it's set to zero?

**Comment by ksooo** [comment]
Even worse, it might crash. As the (too small) struct gets created by the addon and then will be passed to Kodi, which assumes another size of the struct, reading iPosition by Kodi will be out of struct bound with unpredictable results! 

Thus, we're facing an incompatible API charge here. Means, we must bump pvr API min version which means all addons need to be bumped.

**Comment by phunkyfish** [comment]
Bummer.  So how does that work. Do I create PRs that don’t build in Jenkins first for each addon adding the right fields. Then merge the above PR and rebuild each one.

**Comment by ksooo** [comment]
> Do I create PRs that don’t build in Jenkins first for each addon 

 yes, one PR for every official pvr addon => get the list of all official pvrs addon from https://github.com/xbmc/repo-binary-addons - Matrix branch

> adding the right fields. Then merge the above PR and rebuild each one.

 That's not nessecary. Just do a bump of the addon version and update the changelog ("Recompile for 6.1.0 PVR Addon API compatibility" or the like). 

 Welcome to the world of incompatible PVR Addon API changes. :-)

**Comment by phunkyfish** [comment]
That’s not error prone at all...

**Comment by phunkyfish** [comment]
Ok, I’m on it.

**Comment by phunkyfish** [comment]
As it's an API change I assume it's a bump in minor version for each addon?

**Comment by ksooo** [comment]
> As it's a minor API change I assume it's a bump in minor version for each addon?

 Yep.

**Comment by phunkyfish** [comment]
Ok, all PRs submitted. Ok to start by merging this?

**Comment by phunkyfish** [comment]
Sorry, I only just saw your follow up replies on the PRs, I should have added that in hindsight.

**Comment by ksooo** [comment]
> Ok to start by merging this?

 Let me do this as I have commit rights for all affected repos. Okay?

**Comment by ksooo** [comment]
Next time, please put a link to the "causing" Kodi PR in the addon PRs. This way, we would have gotten a nice list of all affected addons in this PR.

If you like, you could do that even now.

**Comment by phunkyfish** [comment]
Sure, I’ll do that now. Gimme 10 mins.

**Comment by ksooo** [comment]
Cool, thx.

**Comment by phunkyfish** [comment]
Ready

**Review by ksooo — APPROVED** [review]
Here we go. :-)

**Comment by Rechi** [comment]
@ksooo please always set milestone before or right after merging a PR.

**Comment by phunkyfish** [comment]
Woohoo!

**Comment by ksooo** [comment]
Done.

**Comment by phunkyfish** [comment]
Nice work ;)

At some point I'll write a script to automate this process, might make it a little less painless.

**Comment by ksooo** [comment]
That's definitely a good idea.

**Pull request merged**
